src/hw/driver/virt_serial.c

```diff
@@ -0,0 +1,22 @@
+#include "common/serial_port.h"
+#include <stdint.h>
+
+SerialPrint platform_serial;
+
+static volatile uint32_t* const UART0 = (uint32_t*)UART_BASE;
+
+static void virt_serial_init() {
+  // In QEMU, we don't have to initialize UART
+}
+
+static void virt_serial_putchar(int c) {
+  *UART0 = (uint32_t)c;
+}
+
+void register_serial(void) {
+  SerialPrint virt = {
+      .init = virt_serial_init,
+      .putchar = virt_serial_putchar,
+  };
+  platform_serial = virt;
+}
```

Comment:
I was able to simplify this by doing:
```
SerialPrint platform_serial = {
 .init = virt_serial_init,
 .putchar = virt_serial_putchar,
};
```
(after you define the functions obviously) Instead of having a register function we have to call, the global will just get init automatically.